connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java

```diff
@@ -753,6 +678,146 @@ private ConnectorsAndTasks assignment(Map<String, ConnectorsAndTasks> memberAssi
+     *                that are already due to be revoked from the worker in this rebalance
+     * @return which connectors and tasks should be revoked from which workers; never null, but may be empty
+     * if no load-balancing revocations are necessary or possible
+     */
+    private Map<String, ConnectorsAndTasks> performLoadBalancingRevocations(
+            ConnectorsAndTasks configured,
+            Collection<WorkerLoad> workers
+    ) {
+        if (log.isTraceEnabled()) {
+            workers.forEach(wl -> log.trace(
+                    "Per worker current load size; worker: {} connectors: {} tasks: {}",
+                    wl.worker(), wl.connectorsSize(), wl.tasksSize()));
+        }
+
+        if (workers.stream().allMatch(WorkerLoad::isEmpty)) {
+            log.trace("No load-balancing revocations required; all workers are either new "
+                    + "or will have all currently-assigned connectors and tasks revoked during this round"
+            );
+            return Collections.emptyMap();
+        }
+        if (configured.isEmpty()) {
+            log.trace("No load-balancing revocations required; no connectors are currently configured on this cluster");
+            return Collections.emptyMap();
+        }
+
+        Map<String, ConnectorsAndTasks.Builder> result = new HashMap<>();
+
+        Map<String, Set<String>> connectorRevocations = loadBalancingRevocations(
+                "connector",
+                configured.connectors().size(),
+                workers,
+                WorkerLoad::connectors
+        );
+        Map<String, Set<ConnectorTaskId>> taskRevocations = loadBalancingRevocations(
+                "task",
+                configured.tasks().size(),
+                workers,
+                WorkerLoad::tasks
+        );
+
+        connectorRevocations.forEach((worker, revoked) ->
+                result.computeIfAbsent(worker, w -> new ConnectorsAndTasks.Builder()).addConnectors(revoked)
+        );
+        taskRevocations.forEach((worker, revoked) ->
+                result.computeIfAbsent(worker, w -> new ConnectorsAndTasks.Builder()).addTasks(revoked)
+        );
+
+        return buildAll(result);
+    }
+
+    private <E> Map<String, Set<E>> loadBalancingRevocations(
+            String allocatedResourceName,
+            int totalToAllocate,
+            Collection<WorkerLoad> workers,
+            Function<WorkerLoad, Collection<E>> workerAllocation
+    ) {
+        int totalWorkers = workers.size();
+        // The minimum instances of this resource that should be assigned to each worker
+        int minAllocatedPerWorker = totalToAllocate / totalWorkers;
+        // How many workers are going to have to be allocated exactly one extra instance
+        // (since the total number to allocate may not be a perfect multiple of the number of workers)
+        int extrasToAllocate = totalToAllocate % totalWorkers;
```

Comment:
i think this could be called `workersWithExtraLoad` to signify `How many workers are going to have to be allocated exactly one extra instance`?

Reply:
I can see how `extrasToAllocate` is misleading, but I'm a little hesitant to use `workersWithExtraLoad` since it could be interpreted as "the number of workers that are running too many connectors/tasks".

Thoughts on `workersToAllocateExtra`? It's a little verbose but should be fairly clear.

Reply:
`workersToAllocateExtra` sounds good to me.